sled-agent/src/artifact_store.rs

```diff
@@ -635,141 +609,129 @@ impl DatasetsManager for StorageHandle {
 /// Abstraction that handles writing to several temporary files.
 #[derive(Debug)]
 struct ArtifactWriter {
-    datasets: usize,
-    files: Vec<Option<(NamedUtf8TempFile<File>, Utf8PathBuf)>>,
-    hasher: Sha256,
+    senders: Vec<mpsc::Sender<Bytes>>,
+    write_tasks: JoinSet<Result<(), Error>>,
     log: Logger,
     sha256: ArtifactHash,
 }
 
 impl ArtifactWriter {
-    /// Calls [`ArtifactWriter::write`] for each chunk in the stream, then
-    /// [`ArtifactWriter::finalize`]. See the documentation for these functions
-    /// for error handling information.
-    async fn write_stream(
-        self,
-        stream: impl Stream<Item = Result<impl AsRef<[u8]>, Error>>,
-    ) -> Result<ArtifactPutResponse, Error> {
-        let writer = stream
-            .try_fold(self, |mut writer, chunk| async {
-                writer.write(chunk).await?;
-                Ok(writer)
-            })
-            .await?;
-        writer.finalize().await
+    fn new(log: Logger, sha256: ArtifactHash) -> ArtifactWriter {
+        ArtifactWriter {
+            senders: Vec::new(),
+            write_tasks: JoinSet::new(),
+            log,
+            sha256,
+        }
     }
 
-    /// Write `chunk` to all temporary files.
-    ///
-    /// Errors in this method are considered non-fatal errors. All errors
-    /// are logged. If all files have failed, this method returns the most
-    /// recently-seen non-fatal error as a fatal error.
-    async fn write(&mut self, chunk: impl AsRef<[u8]>) -> Result<(), Error> {
-        self.hasher.update(&chunk);
-
-        let mut last_error = None;
-        for option in &mut self.files {
-            if let Some((mut file, mountpoint)) = option.take() {
-                match file.as_file_mut().write_all(chunk.as_ref()).await {
-                    Ok(()) => {
-                        *option = Some((file, mountpoint));
+    fn add_path(&mut self, mountpoint: Utf8PathBuf, temp_dir: Utf8PathBuf) {
+        let log = self.log.clone();
+        let path = mountpoint.join(self.sha256.to_string());
+        let atomic_file = AtomicFile::new_with_tmpdir(
+            &path,
+            OverwriteBehavior::AllowOverwrite,
+            temp_dir,
+        );
+        let (tx, mut rx) = mpsc::channel(16); // TODO
```

Comment:
Is there something in particular you plan to do for this `// TODO`?  If it's just that this seems like an arbitrary number, that's fair -- I'd just document that.

It's unfortunate that the buffering here is per-message and not based on the amount of bytes.  I don't really see an easy way around that though and I'm not sure it's worth doing much work to improve it.  Maybe some day it'll be worth creating an async atomicwrites.

Reply:
It's maybe worth noting that this is the number of chunks from dropshot's `StreamingBody` we're willing to buffer up if the writing task(s) are slow? I'm not sure how big those chunks are in practice, nor how likely a slow writer is.

Reply:
When I was implementing it, the typical chunk sizes I saw were around 50-100KiB.

Reply:
I meant to figure out what the expected chunk sizes are (we have both dropshot's `StreamingBody` and reqwest's `Body` to take into account here) and then go from there but haven't yet.

Reply:
`StreamingBody` just forwards the `Bytes` chunks returned from reqwest's `Body` -- there's no rechunking or additional copies.

Reply:
I think a max mem use of up to 64 MB is fine, so 64 (assuming the 1 MB upper bound per chunk) seems reasonable.